rust/src/io/exec/planner.rs

```diff
@@ -0,0 +1,275 @@
+    fn number(&self, value: &str) -> Result<Expr> {
+        use datafusion::logical_expr::lit;
+        if let Ok(n) = value.parse::<i64>() {
+            Ok(lit(n))
+        } else {
+            value
+                .parse::<f64>()
+                .map(lit)
+                .map_err(|_| Error::IO(format!("'{value}' is not supported number value.")))
+        }
+    }
+
+    fn value(&self, value: &Value) -> Result<Expr> {
+        use datafusion::scalar::ScalarValue;
+
+        Ok(match value {
+            Value::Number(v, _) => self.number(v.as_str())?,
+            Value::SingleQuotedString(s) => Expr::Literal(ScalarValue::Utf8(Some(s.clone()))),
+            Value::DollarQuotedString(s) => todo!(),
+            Value::EscapedStringLiteral(_) => todo!(),
+            Value::NationalStringLiteral(_) => todo!(),
+            Value::HexStringLiteral(_) => todo!(),
+            Value::DoubleQuotedString(_) => todo!(),
+            Value::Boolean(v) => Expr::Literal(ScalarValue::Boolean(Some(*v))),
+            Value::Null => Expr::Literal(ScalarValue::Null),
+            Value::Placeholder(_) => todo!(),
+            Value::UnQuotedString(_) => todo!(),
+        })
+    }
+
+    /// True value
+    ///
+    /// For the expression that Lance does not support, just return True, and
+    /// let the upper runtime engine to filter.
+    fn true_value(&self) -> Result<Expr> {
+        Ok(Expr::Literal(ScalarValue::Boolean(Some(true))))
+    }
+
+    fn parse_sql_expr(&self, expr: &SQLExpr) -> Result<Expr> {
+        match expr {
+            SQLExpr::Identifier(id) => self.column(vec![id.clone()].as_slice()),
+            SQLExpr::CompoundIdentifier(ids) => self.column(ids.as_slice()),
+            SQLExpr::BinaryOp { left, op, right } => self.binary_expr(left, op, right),
+            SQLExpr::Value(value) => self.value(value),
+            SQLExpr::IsFalse(expr) => Ok(Expr::IsFalse(Box::new(self.parse_sql_expr(expr)?))),
+            SQLExpr::IsNotFalse(_) => Ok(Expr::IsNotFalse(Box::new(self.parse_sql_expr(expr)?))),
+            SQLExpr::IsTrue(expr) => Ok(Expr::IsTrue(Box::new(self.parse_sql_expr(expr)?))),
+            sqlparser::ast::Expr::IsNotTrue(expr) => {
+                Ok(Expr::IsNotTrue(Box::new(self.parse_sql_expr(expr)?)))
+            }
+            sqlparser::ast::Expr::IsNull(expr) => {
+                Ok(Expr::IsNull(Box::new(self.parse_sql_expr(expr)?)))
+            }
+            sqlparser::ast::Expr::IsNotNull(_) => {
+                Ok(Expr::IsNotNull(Box::new(self.parse_sql_expr(expr)?)))
+            }
+            _ => {
+                return Err(Error::IO(format!(
+                    "Expression '{expr}' is not supported as filter in lance"
+                )))
+            }
+        }
+    }
+
+    /// Create Logical [Expr] from a SQL filter clause.
+    pub fn parse_filter(&self, filter: &str) -> Result<Expr> {
+        let sql = format!("SELECT 1 FROM t WHERE {filter}");
```

Comment:
what is this?

Reply:
This is trick sqlparser to parse a SQL statement, and pick the `WHERE` clause from the statement

Reply:
worth adding a code comment for posterity